### PR TITLE
[irods404] Fix password_obfuscation in Windows

### PR DESCRIPTION
**Pull request by jpmcfarland**

### CODE REVIEW

File: irods/password_obfuscation.py
Comment (d-w-moore) [inline irods/password_obfuscation.py:63]: If we were ever interested in a slightly improved cryptographic strength, we could always use something like hash(os.getlogin( )) or even a int(hashlib.sha256(blah).hexdigest( ),16)), rather of the sum of the character ord-values.  But hey,  this gets the job done : )

File: irods/password_obfuscation.py
Comment (d-w-moore) [inline irods/password_obfuscation.py:112]: Same comment as above : )

File: irods/password_obfuscation.py
Comment (d-w-moore) [inline irods/password_obfuscation.py:63]: If we were ever interested in a slightly improved cryptographic strength, we could always use something like `hash(os.getlogin( ))` or even a `int(hashlib.sha256(blah).hexdigest( ),16)`, rather of the sum of the character ord-values.  But hey,  this gets the job done : )

**Review by d-w-moore — APPROVED** [review]

File: irods/password_obfuscation.py
Comment (d-w-moore) [reply irods/password_obfuscation.py:112]: or below, I guess....

**Comment by d-w-moore** [comment]
Seems to me this will work fine. I don't have the resources for a Windows test at the moment, however.

File: irods/password_obfuscation.py
Comment (jpmcfarland) [reply irods/password_obfuscation.py:63]: In the end, it's just salt for local use.  It need only be consistent.  I don't know what extra strength is really required.

Comment (d-w-moore) [reply irods/password_obfuscation.py:63]: @jpmcfarland Yes, that is definitely enough for the purpose at hand.

Comment (jpmcfarland) [reply irods/password_obfuscation.py:63]: Only one machine so far.  I have asked others if they can test using an editable pip install.  Not too many of us have access to unmanaged Windows machines.

**Comment by d-w-moore** [comment]
One final request @jpmcfarland . Before merging this , I'd at least like to abstract the UID hasher used as a function in `irods.password_obfuscation`.  For my testing purposes as we're looking for a way to allow windows testing.

For example, I'd really like to be able to import it from other modules such as `irods.test.helpers`. 

So maybe we  something like:
```
#(in password_obfuscation.py):
def uid_hash(login_string) : return sum((ord(char) for char in login_string) 
```
and you might patch your code above to say 
```
uid = uid_hash(os.getlogin( )) 
```
in the except clauses.

**Comment by jpmcfarland** [comment]
Sounds good to me!

**Comment by d-w-moore** [comment]
Great, let me know if you'd rather "D.I.Y." or do it via e-mail patch or perhaps pull request to your repo.

**Comment by jpmcfarland** [comment]
I'm modifying it now and will add it to this PR if that's alright.  One question, will this need to be Python 2 compatible (i.e., can I use type hints)?

**Comment by d-w-moore** [comment]
@jpmcfarland Well, so far we've been Python2 compatible; and I feel a patch release ( upcoming v1.1.6) isn't the time to change that.  Do you feel a compelling reason to use type hints?

**Comment by trel** [comment]
Can you rebase (no merge commits)... and squash to a single commit?

**Comment by jpmcfarland** [comment]
@trel Okay.  My IDE refused to push my rebase/squash, then ended up merging another commit.  HELP!

**Comment by trel** [comment]
It doesn't look like you granted permission to push to your branch - so I/we can't help much on this PR...

I'm happy to open another for you with your commits if that is helpful.  (I recommend using a new branch name if you try, so you can keep your main branch out of the conversation, less confusing/troublesome that way.)

**Comment by korydraughn** [comment]
@jpmcfarland What we can do is close this PR and I can create a new one with your changes squashed into one commit. You'll still receive the credit for the work too.

Sound good?

**Comment by jpmcfarland** [comment]
Sure, but I still don't understand completely what's happening on the Git side of things.  And if I don't understand that, I cannot prevent the same thing happening again.  What do I need to do for this, if anything?

**Comment by korydraughn** [comment]
You have merge commits in your git history. Those make it more difficult to rebase.

I think your issues stem from the changes being done on your main branch and merge commits.

A better workflow would be to keep your main branch in sync with upstream and create feature branches off of your main branch.

Basically, you never make changes to your main branch. You always work on a branch created from your main branch. That keeps the rebase process simpler.

**Comment by korydraughn** [comment]
I've created the new PR (#409) with the changes. Please confirm that everything looks as it should.

If everything is correct, then let's close this PR.

**Comment by jpmcfarland** [comment]
> A better workflow would be to keep your main branch in sync with upstream and create feature branches off of your main branch.
> 
> Basically, you never make changes to your main branch. You always work on a branch created from your main branch. That keeps the rebase process simpler.

That's what I normally do.  I don't know why I did it differently here.

Thanks for the help!  Is there any way to "fix" my main branch?  Is deleting/reforking too drastic, or can things work properly after this with or without some adjustment?

**Comment by korydraughn** [comment]
Yes. It can be fixed.

## Solution A
- (optional step) Create a new branch from your main branch (this is your backup branch)
- In your main branch, use `tig` or some other tool to view the commit history
- Starting at the tip of your main branch's history, delete commits until the history is flat and contains no merge commits. You can delete as many commits as you want in this step as long as they are sequential
- Pull the latest changes from upstream/main

## Solution B
- Rename your local main branch
- Check out `<upstream-remote>/main` (this puts you in a detached HEAD state)
- Creating a new local main branch (e.g. `git checkout -b main`)
- Set your local main branch to track `origin/main`
- Pull the latest changes from upstream/main

**Comment by trel** [comment]
Then, in both A and B...
when you push your updated `main` branch, this PR will update and we'll see the new/clean commits.

**Comment by jpmcfarland** [comment]
If I understand this process, until the new commit is merged into upstream/main, it will not be in my fork, only the current main commits (i.e., ending with commit 8311789 at the moment).  Correct?

**Comment by korydraughn** [comment]
That's correct.

**Comment by jpmcfarland** [comment]
And one last question.  Since the `main` "truth" on the remote will be overwritten by these changes to create a new `main` "truth" from the local, a `git push --force` will be required, correct?

**Comment by jpmcfarland** [comment]
Okay, I did:

```bash
git remote add upstream git@github.com:irods/python-irodsclient.git
git fetch upstream
git branch -m main oldmain
git checkout upstream/main
git checkout -b main
```

but trying to set the tracking gives:

```bash
git branch --track main origin/main
fatal: A branch named 'main' already exists.
```

And with force:

```bash
git branch -f --track main origin/main
fatal: Cannot force update the current branch.
```

Then I checkout ANY other "branch" and force set tracking to `origin/main` and end up right back where I started, merge commits and all!  And `git pull upstream main` only gives conflicts because I'm back where I started.  If I stop before trying to set the tracking, the history is pristine.  I'm guessing I should `git push --force` at this point.

But, of course I do not do this.  I look at the status and see that it is only a minor conflict between two of my original commits.  Resolving that and a `git rebase --continue` (yes, I now have `pull.rebase = true` for this checkout) and I have a properly rebased history with a single squashed commit on the end.  Finally I `git push --force` (only ever on my fork) and it appears everything it where it should have been last week.

Please have a look and see if this is indeed the case.

**Comment by trel** [comment]
Oh, I think this might be it!   Will await more votes, but looks pretty good to me.

**Comment by jpmcfarland** [comment]
Thanks EarlyBird!

**Comment by trel** [comment]
noon in basel :)

**Comment by korydraughn** [comment]
It looks correct to me.

Your main branch has a single commit and reports being 1 commit ahead of upstream.

> And one last question. Since the main "truth" on the remote will be overwritten by these changes to create a new main "truth" from the local, a git push --force will be required, correct?

I don't understand the question. What do you mean by **changes to create a new main "truth" from the local**?

My understanding is that once this PR is merged into upstream's main branch, you'll need to run `git fetch upstream` to let git know there are changes. You can then run `git rebase upstream/main` to see if the changes apply cleanly (I expect the changes to apply just fine because the SHA's should match).

**Comment by jpmcfarland** [comment]
The only remaining issue is which PR to merge?  They are both effectively the same now.  I'm happy either way, though it seems this one makes the most sense, has the most direct history.  I leave it up to you.

**Comment by korydraughn** [comment]
We can use this one.

Please mark the remaining comments as resolved and we'll get this PR merged today.

**Pull request merged**
